build.go

```diff
@@ -172,50 +220,123 @@ func runTestCommandLine(pkg string, tags []string) []string {
 	return append(args, pkg)
 }
 
-// RunTest runs the specified package's tests.
-func RunTest(ctx context.Context, pkg string, tags ...string) error {
-	mg.CtxDeps(ctx, BuildTestDep(pkg, tags...))
-
-	return sh.RunV(mg.GoCmd(), runTestCommandLine(pkg, tags)...)
-}
-
-type testRunner struct {
+// TestRunner implements [mg.Fn] to build (as by [BuildTest]) and run the test binary for a package using "go test."
+type TestRunner struct {
 	pkg  string
 	tags []string
 }
 
-func (tr *testRunner) Name() string {
+var _ mg.Fn = &TestRunner{}
+
+// Name implements [mg.Fn].
+func (tr *TestRunner) Name() string {
 	return tr.ID()
 }
 
-func (tr *testRunner) ID() string {
+// ID implements [mg.Fn].
+func (tr *TestRunner) ID() string {
 	return fmt.Sprintf("run-test-%s", tr.pkg)
 }
 
-func (tr *testRunner) Run(ctx context.Context) error {
-	return RunTest(ctx, tr.pkg, tr.tags...)
+// Run implements [mg.Fn]. It runs the package's test with "go test."
+func (tr *TestRunner) Run(ctx context.Context) error {
+	mg.CtxDeps(ctx, BuildTest(tr.pkg, tr.tags...))
+
+	return sh.RunV(mg.GoCmd(), runTestCommandLine(tr.pkg, tr.tags)...)
 }
 
-// RunTestDep returns a [mg.Fn] that will run the tests for the given package, subject to the given build tags.
-func RunTestDep(pkg string, tags ...string) mg.Fn {
-	return &testRunner{pkg, tags}
+// RunTest returns a [mg.Fn] that will run the tests for the given package, subject to the given build tags.
+func RunTest(pkg string, tags ...string) *TestRunner {
+	return &TestRunner{pkg, tags}
 }
 
-// Test runs unit tests.
-func Test(ctx context.Context, tags ...string) error {
-	// It's technically not necessary to build the tests before running them; "go test" will build them anyway.
+// AllGinkgoTestRunner is a [mg.Fn] that identifies all tests in the project and uses Ginkgo to build and run them.
+type AllGinkgoTestRunner struct {
+	AllTestRunner
+	bin string
+}
+
+var _ mg.Fn = &AllGinkgoTestRunner{}
+
+// Run implements [mg.Fn]. It uses Ginkgo to build test binaries for all applicable packages in the project (as by
+// [AllGinkgoTestBuilder], and then uses Ginkgo to run them all.
+func (agtr *AllGinkgoTestRunner) Run(ctx context.Context) error {
+	// It's technically not necessary to build the tests before running them; "ginkgo run" would build them anyway.
 	// However, we specify BuildTests as a dependency so that _all_ the tests get built before _any_ of them start
 	// running. That makes the output cleaner because lengthy test output doesn't push any build failures off the
 	// top of the screen.
-	mg.CtxDeps(ctx, LoadDependencies, BuildTestsDep(tags...))
+	mg.CtxDeps(ctx,
+		LoadDependencies,
+		Install(agtr.bin, "github.com/onsi/ginkgo/v2/ginkgo"),
+		BuildTests(agtr.tags...).UseGinkgo(agtr.bin),
+	)
+	args := []string{
+		"run",
+		"-p",
+		"--timeout", "10s",
+	}
+	args = append(args, formatTags(ginkgoTagOpt, agtr.tags)...)
+	for _, info := range Packages {
+		if !info.HasTest() {
```

Comment:
If `HasTest` is implemented backward, why does this work?

Reply:
All I can conclude is that a previous version of this code worked, but after I changed it, it was being tested anymore. I should come up with a test to include in this repo for it.